### PR TITLE
Feat add filtering custom blocks

### PR DESCRIPTION
**Pull request by omar-bear**
#### Description
The aim is to introduce a full-text search functionality within the custom block library. The search will allow filtering of blocks based on their name, category, and the username of the person who created them.

#### Backend:

1. **Extend API for Custom Blocks Search**
   - File: `personalized-block-service.js`
   - Task:
     - Add a query parameter `searchTerm` in the `getPersonalizedBlocks` method.
     - Implement logic to filter blocks based on the `searchTerm` against block names, categories, and usernames.
     - Note: The username is not directly stored in the `personalizedBlock` collection. A join-like operation (e.g., MongoDB `$lookup`) will be used to fetch the username from the users collection for filtering.

#### Frontend:

2. **Integrate Full-Text Search Logic in Vue Component**
   - File: `personalizedBlocksListComponent.js`
   - Task:
     - Add a data property `searchTerm` to hold the text entered by the user.
     - Add a method `handleSearch` to update `searchTerm` and call `fetchPersonalizedBlocks` with the `searchTerm` as a parameter.
     - Handle empty states: Display a "No results found" message if the search yields no blocks. If `searchTerm` is empty, display a general empty state message.

3. **Automatically Update Displayed Blocks**
   - File: `toolbox.tmpl.html`
   - Task:
     - The `personalizedBlocks` data property will automatically update based on the filtered results from `personalizedBlocksListComponent.js`. No additional logic is required in this file for displaying filtered results.

#### Acceptance Criteria:

- A full-text search field is integrated within `personalizedBlocksListComponent.js`.
- Blocks can be filtered by the `searchTerm`, which checks against block names, categories, and usernames.
- Empty state messages are handled gracefully, with specific messaging for a general empty state and when a `searchTerm` yields no results.

#### Time Estimate:
1 or 2 days

### RESULT

https://github.com/Badsender-com/LePatron.email/assets/80390318/4528d25f-7263-4272-a5c3-ccda5efbe83f

### CODE REVIEW

File: packages/editor/src/js/vue/components/personalized-blocks-list/personalized-blocks-list-component.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 const Vue = require('vue/dist/vue.common');
 const axios = require('axios');
 const { getPersonalizedBlocks } = require('../../utils/apis');
+const debounce = require('lodash.debounce');
+const DEBOUNCE_DELAY_MS = 400;
```
Comment (FlorianGille) [inline packages/editor/src/js/vue/components/personalized-blocks-list/personalized-blocks-list-component.js:5]: You could extract this variable in an utils file for inputs (if it doesn't exist, create a new one)

File: packages/editor/src/js/vue/components/personalized-blocks-list/personalized-blocks-list-component.js [hunk 2]
```diff
@@ -29,16 +37,19 @@ const PersonalizedBlocksListComponent = Vue.component(
...
-          .get(getPersonalizedBlocks(), { params: { groupId } })
+          .get(getPersonalizedBlocks(), {
+            params: { groupId, searchTerm: this.searchTerm },
+          })
           .then((response) => {
             this.vm.personalizedBlocks(
               response.data?.items.map(({ content, ...blockInformation }) => ({
                 ...content,
                 blockInformation,
-                id: "" // This line is crucial. The `viewModel.loadDefaultBlocks` function will override this `id` when the block is added to the email content. Without this placeholder, there could be issues when adding personalized blocks in emails.
```
Comment (FlorianGille) [inline packages/editor/src/js/vue/components/personalized-blocks-list/personalized-blocks-list-component.js:41]: Is it intentional to delete this comment?

Comment (omar-bear) [reply packages/editor/src/js/vue/components/personalized-blocks-list/personalized-blocks-list-component.js:41]: nope my bad !!

File: packages/editor/src/css/style_mosaico_tools.less [hunk 3]
```diff
@@ -740,6 +740,38 @@
...
+  .input-field.search-container {
+    position: relative;
+    margin-bottom: 1rem;
+
+    .search-input {
+      width: 100%;
+      height: 2.3rem;
+      border: 2px solid #ccc; // Border color when not focused
+      border-radius: 4px; // Rounded corners
+      padding: 4px 8px; // Some padding for better appearance
```
Comment (DecampsRenan) [inline packages/editor/src/css/style_mosaico_tools.less:752]: Can you use rem for padding please :) ?

File: packages/editor/src/css/style_mosaico_tools.less [hunk 4]
```diff
@@ -740,6 +740,38 @@
...
+      &:focus {
+        border-width: 2px;
+        border-color: #26a69a; // Border color when focused
+        outline: none; // Remove default outline
+      }
+    }
+
+    label {
+      position: absolute;
+      right: 8px;
```
Comment (DecampsRenan) [inline packages/editor/src/css/style_mosaico_tools.less:764]: rem

File: packages/editor/src/css/style_mosaico_tools.less [hunk 5]
```diff
@@ -836,6 +868,14 @@
     padding: 1.2em;
   }
 
+  /* Style for the search empty state */
+  #personalized-blocks-list .empty-state-search {
+    text-align: center;
+    font-size: 1.2em;
+    color: gray;
+    padding: 1.2em;
```
Comment (DecampsRenan) [inline packages/editor/src/css/style_mosaico_tools.less:876]: not em for spacing (unless you have good reasons ^^), prefer to use rem

Comment (omar-bear) [reply packages/editor/src/css/style_mosaico_tools.less:876]: i took the same padding of the others text messages classes, I fixed all of them now

**Pull request merged**